### PR TITLE
fix(gatsby-recipes): fix prop validation & creation of cache keys during installs

### PR DESCRIPTION
**Pull request by KyleAMathews**
Fix a couple of bugs from the alpha.2 release + restore/improve resource validation errors.

Validation during development:
Screen Shot 2020-08-04 at 4 33 09 PM

Validation when installing:
Screen Shot 2020-08-04 at 10 07 10 PM

### CODE REVIEW

File: packages/gatsby-recipes/src/graphql-server/index.js [hunk 1]
```diff
@@ -11,6 +11,12 @@ const {
 // even if another instance might already be running. This is necessary to ensure the gatsby
 // develop command does not _not_ run the server if the user is running gatsby recipes at the same time.
 export default async (programPath, forceStart) => {
+  // If this env variable is set, we're in dev mode & assume the recipes API was already started
+  // manually.
+  if (process.env.RECIPES_DEV_MODE) {
```
Comment (KyleAMathews) [inline packages/gatsby-recipes/src/graphql-server/index.js:16]: much better way of getting a dev version of the recipes API

File: packages/gatsby-recipes/src/providers/npm/package-json.js [hunk 2]
```diff
@@ -58,18 +59,20 @@ exports.schema = schema
...
   const newState = { ...currentState, [key]: parsedValue }
+  const diff = await getDiff(currentState, newState)
 
   return {
     id: key,
     name,
     currentState: JSON.stringify(currentState, null, 2),
     newState: JSON.stringify(newState, null, 2),
     describe: `Add ${key} to package.json`,
-    diff: ``, // TODO: Make diff
```
Comment (KyleAMathews) [inline packages/gatsby-recipes/src/providers/npm/package-json.js:72]: snuck this in while at it

File: packages/gatsby-recipes/src/renderer/render.js [hunk 3]
```diff
@@ -117,29 +104,21 @@ const validateResource = (resourceName, context, props) => {
   return error
 }
 
-const handleResource = (
-  resourceName,
-  context,
-  props,
-  allResources,
```
Comment (KyleAMathews) [inline packages/gatsby-recipes/src/renderer/render.js:124]: removed a lot of extra props

File: packages/gatsby-recipes/src/renderer/render.js [hunk 4]
```diff
@@ -151,17 +130,26 @@ const handleResource = (
...
-    if (
-      props._key &&
-      !lodash.isEqual(newResource, resourceMap.get(props._key))
-    ) {
-      resourceMap.set(props._key, newResource)
-      setResources([...resourceMap.values()])
+    if (!lodash.isEqual(newResource, resourceMap.get(trueKey))) {
+      resourceMap.set(trueKey, newResource)
+      // TODO Do we need this? It's causing infinite loops
+      // setResources([...resourceMap.values()])
```
Comment (KyleAMathews) [inline packages/gatsby-recipes/src/renderer/render.js:136]: I think this is for nested resources... maybe @johno remember? Anyways, we'll figure it out again when we return to that.

File: packages/gatsby-recipes/src/renderer/render.js [hunk 5]
```diff
@@ -117,29 +104,21 @@ const validateResource = (resourceName, context, props) => {
...
+  // Initialize
   const { mode } = context
 
-  let key
+  const trueKey = props._key ? props._key : context._uuid
+
+  let cacheKey
   // Only run apply once per resource
   if (mode === `apply`) {
-    key = mode + ` ` + resourceName + ` ` + props._key
```
Comment (KyleAMathews) [inline packages/gatsby-recipes/src/renderer/render.js:138]: some resources have a key & others have auto-generated UUIDs — we weren't reflecting that which meant that all subsequent resources hit the same cache key as the first of its type.

File: packages/gatsby-recipes/src/renderer/render.js [hunk 6]
```diff
@@ -227,18 +215,7 @@ const render = async (recipe, cb, inputs = {}, isApply) => {
 
     RecipesReconciler.render(recipeWithWrapper, plan)
 
-    if (errors.length) {
```
Comment (KyleAMathews) [inline packages/gatsby-recipes/src/renderer/render.js:230]: removed generic error handling for what we actually do have support for right now namely validation.

File: packages/gatsby-recipes/src/providers/npm/package-json.js [hunk 7]
```diff
@@ -58,18 +59,20 @@ exports.schema = schema
 exports.validate = validate
 
 module.exports.plan = async ({ root }, { id, name, value }) => {
+  // let parsedValue = value
```
Comment (graysonhicks) [inline packages/gatsby-recipes/src/providers/npm/package-json.js:62]: Not sure if this comment is meant to still be in.

Comment (KyleAMathews) [reply packages/gatsby-recipes/src/providers/npm/package-json.js]: nope! Thanks!

File: packages/gatsby-recipes/src/recipe-machine/index.js [hunk 8]
```diff
@@ -148,7 +148,16 @@ const recipeMachine = Machine(
...
+          CONTINUE: [
+            {
+              target: `applyingPlan`,
+              cond: `hasNoErrors`,
+            },
+            {
+              target: `doneError`,
+              cond: `hasErrors`,
+            },
+          ],
```
Comment (wardpeet) [inline packages/gatsby-recipes/src/recipe-machine/index.js:160]: Technically this isn't necessary, xstate will try the first guard if it's false it goes to the next. They don't run serially or in parallel.
```suggestion
          CONTINUE: [
            {
              target: `doneError`,
              cond: `hasErrors`,
            },
            {
              target: `applyingPlan`,
            },
          ],
```

File: packages/gatsby-recipes/src/cli/index.js [hunk 9]
```diff
@@ -542,16 +550,25 @@ export default async ({
...
-        return <Error width="100%" state={state} />
+        process.nextTick(() => process.exit())
+        return (
+          <ResourceProvider
+            value={
+              state.context.plan?.filter(p => p.resourceName !== `Input`) || []
+            }
+          >
+            <Text bold>
+              Your recipe didn't validate. Please fix the following errors:
```
[the remark this answers is not in the document]
Comment (muescha) [reply packages/gatsby-recipes/src/cli/index.js:562]: Maybe replace also "your" - if the recipe are from an other person so the user get not "blamed" for it?

Just "this recipe"

[the remark this answers is not in the document]
Comment (wardpeet) [reply packages/gatsby-recipes/src/cli/index.js:562]: thanks!

File: packages/gatsby-recipes/src/graphql-server/index.js (hunk 1 above)
Comment (muescha) [reply packages/gatsby-recipes/src/graphql-server/index.js:16]: Maybe check here if the server is running and print a error if not

File: packages/gatsby-recipes/README.md [hunk 10]
```diff
@@ -241,7 +241,7 @@ DEBUG=true node --inspect-brk ./node_modules/.bin/gatsby recipes ./test.mdx
 
 Then, open up Chrome and click the node icon in dev tools.
```
Comment (muescha) [inline packages/gatsby-recipes/README.md:242]: add info here about the `./recipe-server.log` with `DEBUG=true`?

File: packages/gatsby-recipes/src/cli/index.js [hunk 11]
```diff
@@ -542,16 +550,25 @@ export default async ({
...
-        if (state && state.context && state.context.error) {
-          return <Text red>{JSON.stringify(state.context.error, null, 2)}</Text>
-        }
-
-        return null
-      }
-
       if (state?.value === `doneError`) {
-        return <Error width="100%" state={state} />
+        process.nextTick(() => process.exit())
```
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/gatsby-recipes/src/cli/index.js:554]: Ink v2 wouldn't exit cleanly under some conditions. Perhaps v3 is better.

File: packages/gatsby-recipes/src/cli/index.js [hunk 12]
```diff
@@ -542,16 +550,25 @@ export default async ({
...
+        return (
+          <ResourceProvider
+            value={
+              state.context.plan?.filter(p => p.resourceName !== `Input`) || []
```
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/gatsby-recipes/src/cli/index.js:558]: inputs are components too like resources which is why we exclude them. The input from the inputs (haha) are ignored unless they're passed as props into a resource component in which case they're validated like normal.

File: packages/gatsby-recipes/src/renderer/render.js (hunk 4 above)
Comment (johno) [reply packages/gatsby-recipes/src/renderer/render.js:136]: I think at one point it was for nested resources, but the code that was leveraging it is long gone IIRC.

File: packages/gatsby-recipes/src/cli/index.js [hunk 13]
```diff
@@ -542,16 +550,25 @@ export default async ({
...
+            }
+          >
+            <Text bold>
+              Your recipe didn't validate. Please fix the following errors:
+            </Text>
+            <Text>{`\n`}</Text>
+            {state.context.plan
+              .filter(p => p.error)
+              .map((p, i) => (
+                <ResourceComponent key={i} {...p} />
```
[the remark this answers is not in the document]
Comment (johno) [reply packages/gatsby-recipes/src/cli/index.js:568]: The resource should have a `_uuid` which we can use here instead.

File: packages/gatsby-recipes/src/cli/index.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/gatsby-recipes/src/cli/index.js:558]: adding it as a comment

File: packages/gatsby-recipes/src/cli/index.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/gatsby-recipes/src/cli/index.js:562]: 👍 to /s/your/this

File: packages/gatsby-recipes/src/cli/index.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/gatsby-recipes/src/cli/index.js:568]: this renders and exits so it's pretty safe :-D — it's just to keep React quiet.

File: packages/gatsby-recipes/src/providers/npm/package-json.js [hunk 14]
```diff
@@ -62,14 +63,15 @@ module.exports.plan = async ({ root }, { id, name, value }) => {
   const key = id || name
```
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/gatsby-recipes/src/providers/npm/package-json.js:63]: oh right — we just need to use `value` now

File: packages/gatsby-recipes/src/providers/npm/package-json.js [hunk 15]
```diff
@@ -48,7 +49,7 @@ const destroy = async ({ root }, { id }) => {
 
 const schema = {
   name: Joi.string(),
-  value: Joi.string(),
+  value: Joi.object(),
```
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/gatsby-recipes/src/providers/npm/package-json.js:52]: it is but we're in alpha so breaking changes are expected. Also we hadn't documented this resource yet for this exact reason as we knew we wanted it to work this way.

File: packages/gatsby-recipes/src/recipe-machine/index.js (hunk 8 above)
Comment (KyleAMathews) [reply packages/gatsby-recipes/src/recipe-machine/index.js:160]: 👍 good idea

File: packages/gatsby-recipes/README.md (hunk 10 above)
Comment (KyleAMathews) [reply packages/gatsby-recipes/README.md:242]: we've moved away from `recipe-server.log` as it's a lot easier to just run the API process in its own terminal window + it's tricky to make sure everything is logged when running things in the background.

**Pull request merged**